File: src/functions/numeric/python/interpolation/Newton.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Computational Solutions                                      April 22, 2022
IST 4360
Prof. M Diaz Maldonado

Synopsis:
Uses a Newton 3rd-degree polynomial to interpolate a dataset (xi, yi):

    y = a0 + a1(x - x1) + a2(x - x1)(x - x2) + a3(x - x1)(x - x2)(x - x3),

where y = f(x), and the a's are the coefficients of the polynomial.


Copyright (c) 2022 Misael Diaz-Maldonado
This file is released under the GNU General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.


References:
[0] A Gilat and V Subramanian, Numerical Methods for Engineers and
    Scientists: An Introduction with Applications using MATLAB
[1] R Johansson, Numerical Python: Scientific Computing and Data
    Science Applications with NumPy, SciPy, and Matplotlib, 2nd edition
"""


from numpy import abs
from numpy import array
from numpy import linspace
from numpy.linalg import det
from numpy.linalg import cond
from numpy.linalg import solve
import matplotlib as mpl
mpl.use('qt5agg')
from matplotlib import pyplot as plt
# imports from user-defined modules
from nlsolvers import bisect, nls_opts
from nint import trap


""" Dataset """
# xi, yi coordinates
x1, y1 = -1.2,  8.8
x2, y2 =  0.2, -5.0
x3, y3 =  2.0,  6.0
x4, y4 =  3.5,  5.0


""" Newton's 3rd-degree interpolating polynomial """
# differences
D0, D1 = y1, (y2 - y1) / (x2 - x1)
D2, D3 = (y3 - y2) / (x3 - x2), (y4 - y3) / (x4 - x3)
# coefficients
a0, a1 = D0, D1
a2 = (D2 - D1) / (x3 - x1)
a3 = ( (D3 - D2) / (x4 - x2) - (D2 - D1) / (x3 - x1) ) / (x4 - x1)
# defines the 3rd degree polynomial as a lambda (or anonymous) function
f = lambda x: (
    a0 + a1 * (x - x1) + a2 * (x - x1) * (x - x2) +
    a3 * (x - x1) * (x - x2)* (x - x3)
)


""" post-processing """
# computes the differences between the data and the approximation returned
# by the interpolating function f(x)
diffs = (
    f'\nDifferences:\n'
    f'{abs( y1 - f(x1) ):.12f}\n'
    f'{abs( y2 - f(x2) ):.12f}\n'
    f'{abs( y3 - f(x3) ):.12f}\n'
    f'{abs( y4 - f(x4) ):.12f}\n'
)
# should display an array of zeros to indicate success
print(diffs)


# as above but accumulates the differences
xi = array([x1, x2, x3, x4])
yi = array([y1, y2, y3, y4])
diffs = (
    f'\nCumulative Difference:\n'
    f'{abs( yi - f(xi) ).sum():.12f}\n'
)
# should display a zero (scalar) to indicate success
print(diffs)


""" plots the dataset and the interpolating polynomial """
plt.close('all')
plt.ion()
fig, ax = plt.subplots()
# plots the x, y coordinates
ax.plot(xi, yi,  color="red",   marker="*", markersize=12, linestyle="")
ax.set_xlim([ -2,  5])
ax.set_ylim([-30, 50])
# plots the interpolating polynomial
x = linspace(-2, 5, 256)
ax.plot(x, f(x), color="black", linestyle="--")
ax.set_xlabel('x')
ax.set_ylabel('f(x)')


""" Solves for the largest root of f(x): """

# copies nonlinear solver options into a dictionary
options = nls_opts
# overrides defaults
options['tol']      = 1.0e-12   # tolerance
options['max_iter'] = 256       # maximum number of iterations
options['verbose']  = True      # enables report

lims = (3, 5)
r = bisect(lims, f, opts = options)
print(f'largest root of f(x): {r:.15f}')


""" Integrates the function f(x) in the specified range: """

N = 256                 # number of integration intervals
lb, ub = (-1.2, 3.5)    # lower and upper integration limits
# applies the trapezoid method
integral = trap(lb, ub, N, f)
print(f'area under the curve: {integral:.15f}')

File: src/functions/numeric/matlab/interpolation/Newton.m
% Computational Solutions                                    April 22, 2022
% IST 4360
% Prof. M Diaz Maldonado
%
% Synopsis:
% Uses a Newton 3rd-degree polynomial to interpolate a dataset (xi, yi):
%
%    y = a0 + a1(x - x1) + a2(x - x1)(x - x2) + a3(x - x1)(x - x2)(x - x3),
%
% where y = f(x), and the a's are the coefficients of the polynomial.
%
%
% Copyright (c) 2022 Misael Diaz-Maldonado
% This file is released under the GNU General Public License as published
% by the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
%
% References:
% [0] A Gilat and V Subramanian, Numerical Methods for Engineers and
%     Scientists: An Introduction with Applications using MATLAB
% [1] A Gilat, MATLAB: An Introduction with Applications, 6th edition


clear
close all
clc

format long g


% Dataset:
% xi, yi coordinates
x1 = -1.2; y1 =  8.8;
x2 =  0.2; y2 = -5.0;
x3 =  2.0; y3 =  6.0;
x4 =  3.5; y4 =  5.0;


% Newton's 3rd-degree interpolating polynomial:
% differences
D0 = y1;
D1 = (y2 - y1) / (x2 - x1);
D2 = (y3 - y2) / (x3 - x2);
D3 = (y4 - y3) / (x4 - x3);
% coefficients
a0 = D0;
a1 = D1;
a2 = (D2 - D1) / (x3 - x1);
a3 = ( (D3 - D2) / (x4 - x2) - (D2 - D1) / (x3 - x1) ) / (x4 - x1);
% Note: the anonymous function has been defined for a scalar input
f = @(x) (						...
	a0 + a1 * (x - x1) + a2 * (x - x1) * (x - x2) + ...
	a3 * (x - x1) * (x - x2)* (x - x3)		...
);


% Post-processing:

% computes the differences between the data and the approximation returned
% by the interpolating function f(x)
x = [x1; x2; x3; x4];
y = [y1; y2; y3; y4];

% should display an array of zeros to indicate success
diffs = 0;
fprintf("\nDifferences:\n")
for i = 1:length(x)
	diffs = diffs + abs( y(i) - f( x(i) ) );
	fprintf( "%.12f\n", abs( y(i) - f( x(i) ) ) )
end

% as above but accumulates the differences
fprintf("\nCumulative Difference:\n")
fprintf("%.12f\n", diffs)

% Plots the dataset and the interpolating polynomial:
figure(1)

plot(x, y, 'r+')		% plots the xi, yi coordinates
hold on
% performs an elementwise computation of the interpolating polynomial
x = linspace(-2, 5, 256);
y = zeros( size(x) );
for i = 1:length(x)
	y(i) = f( x(i) );	% Note: the function expects a scalar input
end
plot(x, y, '--k')		% plots interpolating polynomial
xlabel('x')
ylabel('f(x)')


% Solves for the largest root of f(x):

% sets limits of the bracketing interval [lb, ub] enclosing largest root
lb = 3;	% lower bound
ub = 5;	% upper bound
% overrides default tolerance and maximum number of iterations
opt = struct('tol', 1.0e-12, 'max_iter', 256, 'verbose', 1);
% applies the bisection method
r = bisect(lb, ub, f, opt);
fprintf("largest root of f(x): %.15f\n", r);


% Integrates the function f(x) in the specified range:
N = 256;	% number of integration intervals
lb = -1.2;	% lower integration limit
ub =  3.5;	% upper integration limit
% applies the trapezoid method
int = trap(lb, ub, N, f);
fprintf("area under the curve: %.15f\n", int);
